daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala

```diff
@@ -1099,31 +1098,76 @@ private[lf] object Speedy {
             }
           case TTyCon(tyCon) =>
             value match {
-              case V.ValueRecord(_, fields) =>
-                val lookupResult =
-                  assertRight(compiledPackages.pkgInterface.lookupDataRecord(tyCon))
+              case V.ValueRecord(_, sourceElements) => {
+                val lookupResult = assertRight(
+                  compiledPackages.pkgInterface.lookupDataRecord(tyCon)
+                )
+                val targetFieldsAndTypes: ImmArray[(Name, Type)] = lookupResult.dataRecord.fields
                 lazy val subst = lookupResult.subst(argTypes)
-                // TODO: https://github.com/digital-asset/daml/issues/17082
-                // - disallow dropping extra fields which have value other than None.
-                val actualValues = (lookupResult.dataRecord.fields.iterator zip fields.iterator)
-                  .map { case ((_, fieldType), (_, fieldValue)) =>
-                    go(AstUtil.substitute(fieldType, subst), fieldValue)
+
+                // This code implements the compatibility transformation used for up/down-grading
+                // And handles the cases:
+                // - UPGRADE:   numT > numS : creates a None for each missing fields.
+                // - DOWNGRADE: numS > numT : drops each extra field, ensuring it is None.
+                val numS: Int = sourceElements.length
+                val numT: Int = targetFieldsAndTypes.length
+
+                // traverse the sourceElements, "get"ing the corresponding target type
+                // when there is no corresponding type, we must be downgrading, and so we insist the value is None
+                val values0: List[SValue] =
+                  sourceElements.toSeq.zipWithIndex.flatMap { case ((optName, v), i) =>
+                    targetFieldsAndTypes.get(i) match {
+                      case Some(x) => {
+                        val (targetField, targetFieldType): (Name, Type) = x
+                        optName match {
+                          case None => ()
+                          case Some(sourceField) =>
+                            // value is not normalized; check field names match
+                            assert(sourceField == targetField)
+                        }
+                        val typ: Type = AstUtil.substitute(targetFieldType, subst)
+                        val sv: SValue = go(typ, v)
+                        List(sv)
+                      }
+                      case None => { // DOWNGRADE
+                        assert(i >= numT)
```

Comment:
Should we also assert that `numS > numT` - i.e. we are actually within a downgrade context?

Reply:
I don't think this assertion would add anything, 
`i`  is bound by `sourceElements.toSeq.zipWithIndex` and so ranges from 0 to `numS-1`  
So `i >= numT` implies `numS > numT`
I will add a comment to this effect.

Reply:
Got you.

Yes, so when `i >= numT`, we know at leats one of our source indexes outstrips `numT`, and so can infer that `numS > numT`.

Yeah, a comment here would help a great deal. Thanks for that.

Reply:
Actually. What the hell. I will add the assertion you suggested.